File: apps/sidecar/examples/lintu_client.js
/**
 * 灵图 Open API v1 — Node SDK Demo (zero-dep, Node 18+ uses built-in fetch)
 *
 * Usage:
 *   import { LintuClient } from './lintu_client.js'
 *   const lintu = new LintuClient({
 *     baseUrl: 'https://api.your-domain.com',
 *     token:   'lk_live_xxx.your_secret',
 *   })
 *   const stats = await lintu.stats()
 *   console.log(stats)
 *
 * For browser (H5) use, see "H5 集成" in docs/open-api-guide.md — short
 * version: do NOT put the secret in JS shipped to the browser. Run this
 * client in your H5 product's BACKEND, expose a thin proxy to the page.
 */

export class LintuAPIError extends Error {
  constructor(status, body) {
    super(`HTTP ${status}: ${String(body).slice(0, 300)}`)
    this.name = 'LintuAPIError'
    this.status = status
    this.body = body
  }
}

export class LintuClient {
  /**
   * @param {{ baseUrl: string, token: string, timeoutMs?: number }} opts
   */
  constructor({ baseUrl, token, timeoutMs = 30000 }) {
    this.baseUrl = baseUrl.replace(/\/+$/, '') + '/open-api/v1'
    this.token = token
    this.timeoutMs = timeoutMs
  }

  async _request(method, path, { params, body, accept = 'application/json' } = {}) {
    let url = this.baseUrl + path
    if (params) {
      const qs = new URLSearchParams()
      for (const [k, v] of Object.entries(params)) {
        if (v == null) continue
        if (Array.isArray(v)) v.forEach((x) => qs.append(k, String(x)))
        else qs.set(k, String(v))
      }
      const s = qs.toString()
      if (s) url += '?' + s
    }
    const ctl = new AbortController()
    const timer = setTimeout(() => ctl.abort(), this.timeoutMs)
    try {
      const resp = await fetch(url, {
        method,
        signal: ctl.signal,
        headers: {
          Authorization: `Bearer ${this.token}`,
          Accept: accept,
          ...(body ? { 'Content-Type': 'application/json' } : {}),
        },
        body: body ? JSON.stringify(body) : undefined,
      })
      if (!resp.ok) {
        throw new LintuAPIError(resp.status, await resp.text())
      }
      if (accept === 'application/json') return await resp.json()
      return new Uint8Array(await resp.arrayBuffer())
    } finally {
      clearTimeout(timer)
    }
  }

  // ── Endpoints ──

  health() { return this._request('GET', '/health') }
  stats(projectId) { return this._request('GET', '/stats', { params: { project_id: projectId } }) }

  /** @param {{ project_id?: string, source_type?: string, scene?: string[], season?: string[], offset?: number, limit?: number }} q */
  listImages(q = {}) { return this._request('GET', '/images', { params: q }) }

  getImage(id) { return this._request('GET', `/images/${id}`) }
  listDerivatives(id) { return this._request('GET', `/images/${id}/derivatives`) }
  downloadImage(id, { size } = {}) {
    return this._request('GET', `/images/${id}/file`, {
      params: size ? { size } : undefined, accept: 'image/jpeg',
    })
  }

  listTags({ projectId, dimension } = {}) {
    return this._request('GET', '/tags', { params: { project_id: projectId, dimension } })
  }

  matrix(projectId, { row = 'scene', col = 'season' } = {}) {
    return this._request('GET', '/matrix', { params: { project_id: projectId, row, col } })
  }

  listBatches({ projectId, status } = {}) {
    return this._request('GET', '/batches', { params: { project_id: projectId, status } })
  }
  getBatch(id) { return this._request('GET', `/batches/${id}`) }

  submitBatch(body) {
    // body: { project_id, name, task_type, seed_image_ids, prompt_ids, concurrency?, max_retry?, budget_usd? }
    return this._request('POST', '/batches', { body })
  }
}

// ── Demo when run directly ──
if (import.meta.url === `file://${process.argv[1]}`) {
  const base = process.env.LINTU_BASE || 'http://localhost:7879'
  const token = process.env.LINTU_TOKEN
  if (!token) {
    console.error('Set LINTU_TOKEN=lk_live_xxx.secret first')
    process.exit(1)
  }
  const c = new LintuClient({ baseUrl: base, token })
  console.log('health:', await c.health())
  console.log('stats:', await c.stats())
  const images = await c.listImages({ limit: 3 })
  console.log(`got ${images.total} images, first 3:`)
  for (const img of images.items) {
    console.log(`  ${img.id}  ${img.file_name}  ${img.width}x${img.height}`)
  }
}

File: apps/sidecar/examples/lintu_client.py
"""灵图 Open API v1 — Python SDK Demo

Usage::

    from lintu_client import LintuClient

    client = LintuClient(
        base_url="https://api.your-domain.com",
        token="lk_live_xxx.your_secret",
    )

    print(client.health())
    images = client.list_images(limit=10)
    for img in images["items"]:
        print(img["id"], img["file_name"])

    # Submit a batch (requires generate:write scope)
    batch = client.submit_batch(
        project_id="...",
        name="external H5 batch",
        task_type="outpaint",
        seed_image_ids=["img1", "img2"],
        prompt_ids=["p1", "p2"],
        concurrency=5,
        budget_usd=2.0,
    )
    print("batch:", batch["id"])

Dependencies: stdlib only (urllib.request). Drop-in copy is fine.
"""
from __future__ import annotations

import json
import urllib.parse
import urllib.request
from typing import Any, Iterable


class LintuAPIError(Exception):
    def __init__(self, status: int, body: str):
        super().__init__(f"HTTP {status}: {body[:300]}")
        self.status = status
        self.body = body


class LintuClient:
    def __init__(self, base_url: str, token: str, timeout: int = 30):
        self.base_url = base_url.rstrip("/") + "/open-api/v1"
        self.token = token
        self.timeout = timeout

    # ── Internals ──

    def _request(
        self,
        method: str,
        path: str,
        params: dict | None = None,
        body: dict | None = None,
        accept: str = "application/json",
    ):
        url = self.base_url + path
        if params:
            url += "?" + urllib.parse.urlencode(
                {k: v for k, v in params.items() if v is not None},
                doseq=True,
            )
        data = json.dumps(body).encode("utf-8") if body is not None else None
        req = urllib.request.Request(url, data=data, method=method)
        req.add_header("Authorization", f"Bearer {self.token}")
        req.add_header("Accept", accept)
        if data is not None:
            req.add_header("Content-Type", "application/json")
        try:
            with urllib.request.urlopen(req, timeout=self.timeout) as resp:
                payload = resp.read()
                if accept == "application/json":
                    return json.loads(payload)
                return payload
        except urllib.error.HTTPError as e:
            raise LintuAPIError(e.code, e.read().decode("utf-8", "replace")) from e

    # ── Endpoints ──

    def health(self) -> dict:
        return self._request("GET", "/health")

    def stats(self, project_id: str | None = None) -> dict:
        return self._request("GET", "/stats", params={"project_id": project_id})

    def list_images(
        self,
        *,
        project_id: str | None = None,
        source_type: str | None = None,
        scene: Iterable[str] | None = None,
        season: Iterable[str] | None = None,
        offset: int = 0,
        limit: int = 50,
    ) -> dict:
        return self._request("GET", "/images", params={
            "project_id": project_id,
            "source_type": source_type,
            "scene": list(scene) if scene else None,
            "season": list(season) if season else None,
            "offset": offset,
            "limit": limit,
        })

    def get_image(self, image_id: str) -> dict:
        return self._request("GET", f"/images/{image_id}")

    def list_derivatives(self, image_id: str) -> dict:
        return self._request("GET", f"/images/{image_id}/derivatives")

    def download_image(self, image_id: str, *, size: int | None = None) -> bytes:
        return self._request("GET", f"/images/{image_id}/file",
                             params={"size": size} if size else None,
                             accept="image/jpeg")

    def list_tags(self, *, project_id: str | None = None, dimension: str | None = None) -> list[dict]:
        return self._request("GET", "/tags", params={
            "project_id": project_id, "dimension": dimension,
        })

    def matrix(self, project_id: str, row: str = "scene", col: str = "season") -> dict:
        return self._request("GET", "/matrix",
                             params={"project_id": project_id, "row": row, "col": col})

    def list_batches(self, *, project_id: str | None = None, status: str | None = None) -> list[dict]:
        return self._request("GET", "/batches", params={
            "project_id": project_id, "status": status,
        })

    def get_batch(self, batch_id: str) -> dict:
        return self._request("GET", f"/batches/{batch_id}")

    def submit_batch(
        self,
        *,
        project_id: str,
        name: str,
        task_type: str,
        seed_image_ids: list[str],
        prompt_ids: list[str],
        concurrency: int = 10,
        max_retry: int = 3,
        budget_usd: float | None = None,
    ) -> dict:
        return self._request("POST", "/batches", body={
            "project_id": project_id,
            "name": name,
            "task_type": task_type,
            "seed_image_ids": seed_image_ids,
            "prompt_ids": prompt_ids,
            "concurrency": concurrency,
            "max_retry": max_retry,
            "budget_usd": budget_usd,
        })


if __name__ == "__main__":
    import os
    import sys

    base = os.environ.get("LINTU_BASE", "http://localhost:7879")
    token = os.environ.get("LINTU_TOKEN")
    if not token:
        print("Set LINTU_TOKEN=lk_live_xxx.secret first", file=sys.stderr)
        sys.exit(1)

    c = LintuClient(base, token)
    print("health:", c.health())
    print("stats:", c.stats())
    images = c.list_images(limit=3)
    print(f"got {images['total']} images, first 3:")
    for img in images["items"]:
        print(f"  {img['id']}  {img['file_name']}  {img['width']}x{img['height']}")
